File: Data Structures & Algorithms/count-number-of-islands/submission-3.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        vector<vector<int>> directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        int count = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int k = 0; k < grid[0].size(); k++) {
                if (grid[i][k] == '1') {
                    dfs(grid, i, k, directions);
                    count++;
                }
            }
        }
        return count;
        


    }

private:
    void dfs(vector<vector<char>>& grid, int r, int c, vector<vector<int>> directions) {
        grid[r][c] = 0;
        for (const auto& dir : directions) {
            int newR = r + dir[0];
            int newC = c + dir[1];
            if (newR >= 0 && newR < grid.size() && newC >= 0 && newC < grid[0].size()) {
                if (grid[newR][newC] == '1') {
                    dfs(grid, newR, newC, directions);
                }
            }
        }
    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-3.java
public class Solution {
    public int orangesRotting(int[][] grid) {
        Queue<int[]> q = new ArrayDeque<>(); //Queue of Rotten Oranges
        int fresh = 0;
        int time = 0;
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                // Counts the number of fresh fruit.
                if (grid[r][c] == 1) { //Increments Fresh Fruit
                    fresh++;
                }
                if (grid[r][c] == 2) { //Adds Rotten Oranges To Queue
                    q.offer(new int[]{r, c});
                }
            }
        }
        // BFS begins.
        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        // Iterate until no fresh fruits remain or if this is impossible.
        while (fresh > 0 && !q.isEmpty()) { 
            int length = q.size();
            for (int i = 0; i < length; i++) {
                // Get coordinates of current rotten orange.
                int[] curr = q.poll();
                int r = curr[0];
                int c = curr[1];
                for (int[] dir : directions) {
                    int row = r + dir[0];
                    int col = c + dir[1];
                    if (row >= 0 && row < grid.length && 
                        col >= 0 && col < grid[0].length &&
                        grid[row][col] == 1) { //Rotten any adjacent oranges.
                        grid[row][col] = 2;
                        // Add the new rotten orange to the queue.
                        q.offer(new int[]{row, col});
                        // Since it's no longer fresh, decrement the count.
                        fresh--;
                    }
                }
            }
            time++;
        }
        return fresh == 0 ? time : -1;
    }
}
